Try a new hover and select approach to improve nested block selection

jasmussen opened the pull request:
This PR aims to solve a number of issues we've had around selecting nested and parent blocks. Watch the GIF first:

![nested blocks](https://user-images.githubusercontent.com/1204802/40106451-1b510e10-58f6-11e8-8341-0de4d5c85079.gif)

It also tries to take the vinegar out of the hover label, and makes it feel much less fiddly to work with blocks in nested situations.

It does this by making the parent block side padding much wider than the nested block padding. *But only in a visual way, using negative margins on the UI*. That means the appearance of the blocks should not change visually, the goal is still to be 1:1 with the front-end, and further enhancements to nested block paddings are coming in #6408.

For starters, please try out this branch and get a feel for it. Depending on feedback and aside from polish, there are two remaining to-do's for this branch:

1. I'd like help with the mobile solution, which is that you click a block with children and select the parent first, and _then_ click again to select or set the caret in a child block.
2. We should look at tuning the keyboard navigation so when you arrow-key down from a paragraph and into a block that has children, you first select the block itself, and then navigate into the child blocks. 

Your thoughts and feedback are appreciated.

Commit:
Alternate hover approach WIP

youknowriad committed:
Show outline on parent blocks

jasmussen commented:
This has been rebased and squashed, so the quote block no longer uses nested blocks like it does in the GIF above. Please test using the Columns block.

afercia reviewed (commented):
@jasmussen thanks for the ping. Tested with columns and two quotes. First impression is that is not immediately clear when/why the "breadcrumb" labels appear and when not. They appear on hover, but not when the block is selected, that's a bit confusing.

Aside: I guess the breadcrumb labels were introduced specifically for the nested block but they also reveal that sometimes there's the need to identify what a block is. They're basically visible labels, I wonder if it could be beneficial to always show them, at least when the block is selected.

The color contrast of the breadcrumb label is not sufficient. See `.editor-block-list__breadcrumb` has two background CSS properties, the latter prevails and it's `$blue-medium-300` #66C6E4 with white text. The contrast ratio of 1.95:1 https://jdlsn.com/color/?type=hex&color=FFFFFF&color2=66C6E4 should be at least 4.5:1

<img width="742" alt="screen shot 2018-05-16 at 11 47 58" src="https://user-images.githubusercontent.com/1682452/40110695-f4ae7896-5900-11e8-8e55-6ddb4b45f8e7.png">

In some cases, for example when hovering the "horizontal inserter" the label appears:

<img width="735" alt="horiontal inserter nested blocks" src="https://user-images.githubusercontent.com/1682452/40110727-0c83fcac-5901-11e8-9a73-40a912327605.png">

That's potentially very confusing, as users may think the label is related to the inserter.

Keyboard interaction: yes as you pointed out there's the need to improve it. Also, the "Add block" plus icon appear only on focus, which makes them hardly discoverable.

jasmussen commented:
Great review, will try to address your points. Keep it coming.

chrisvanpatten commented:
I think this is a big step in making nested blocks easier to work with.

See also #6224, which isn't explicitly related, but definitely has some crossover in terms of exploring the block interaction UI.

karmatosed reviewed (approved):
This is a huge improvement! Thanks for working on this @jasmussen. Experience and design wise I am going to approve this, noting that the accessibility improvements are already being worked on.

jasmussen commented:
Rebased, squashed, and fixed an issue with full-wide blocks.

aduth commented on editor/components/block-list/block.js:
```diff
@@ -647,6 +653,9 @@ const applyWithSelect = withSelect( ( select, { uid, rootUID } ) => {
 		isSelected,
 		hasFixedToolbar,
+		rootUIDOfRoot: getBlockRootUID( rootUID ),
+		orderOfRoot: getBlockIndex( rootUID, getBlockRootUID( rootUID ) ),
```
Unused?

aduth replied:
>Unused?

Removed in 11d7161d16b97b3b0486385bbc46c123586f889d.

aduth commented on editor/components/block-list/block.js:
```diff
@@ -647,6 +653,9 @@ const applyWithSelect = withSelect( ( select, { uid, rootUID } ) => {
 		block,
 		isSelected,
 		hasFixedToolbar,
+		rootUIDOfRoot: getBlockRootUID( rootUID ),
```
Unused?

aduth replied:
>Unused?

Removed in 11d7161d16b97b3b0486385bbc46c123586f889d.

aduth commented on editor/components/block-list/block.js:
```diff
@@ -647,6 +653,9 @@ const applyWithSelect = withSelect( ( select, { uid, rootUID } ) => {
 		hasFixedToolbar,
+		rootUIDOfRoot: getBlockRootUID( rootUID ),
+		orderOfRoot: getBlockIndex( rootUID, getBlockRootUID( rootUID ) ),
+		isSelected,
```
Duplicated (already assigned above). Thinking there ought to be an ESLint rule for this.

_Edit:_ [There is](https://eslint.org/docs/rules/no-dupe-keys) and it's enabled. I'm guessing it doesn't support the ES2015 property shorthand in the version we're using.

aduth replied:
>Duplicated (already assigned above).

Removed in 11d7161d16b97b3b0486385bbc46c123586f889d.

aduth commented on editor/components/block-list/breadcrumb.js:
```diff
@@ -62,14 +61,9 @@ export class BlockBreadcrumb extends Component {
-						/>
+						<BlockTitle uid={ rootUID } />
 					) }
+					{ rootUID && ( <span> &rarr; </span> ) }
```
Directionality has me wondering how this appears on RTL.

aduth replied:
Is the `span` necessary? Could this be `{ rootUID && ' → ' }` ?

_Edit:_ I see we're applying styles. We should probably do so with a class.

aduth replied:
>Directionality has me wondering how this appears on RTL.
>
>[...]
>
>Is the `span` necessary? Could this be `{ rootUID && ' → ' }` ?

Updated to an element with `:after` pseudo-class `content` (and RTL flipping) in e650954c389401426954b87972ea86d1857ff315.

aduth commented on editor/components/block-list/breadcrumb.js:
```diff
@@ -53,7 +52,7 @@ export class BlockBreadcrumb extends Component {
 	}
 
 	render( ) {
-		const { uid, rootUID, selectRootBlock, isHidden } = this.props;
```
If we're not using `selectRootBlock`, the composed `withDispatch` can be removed.

aduth replied:
>If we're not using `selectRootBlock`, the composed `withDispatch` can be removed.

Removed in 11d7161d16b97b3b0486385bbc46c123586f889d.

aduth commented on editor/store/selectors.js:
```diff
@@ -928,6 +929,18 @@ export function isBlockSelected( state, uid ) {
+ *
+ * @return {boolean} Whether the block as an inner block selected
+ */
+export function hasSelectedInnerBlock( state, uid ) {
```
What about an inner multi-selection?

aduth committed:
Block Breadcrumb: Render arrow as RTL compatible

aduth committed:
Writing Flow: Force tab stop at focusable wrapper when inner blocks exist

aduth committed:
Block List: Remove unused variables

aduth committed:
Inner Blocks: Show overlay for small screen, unselected root block

aduth commented:
3a0aac36903e01b9551ace3d6114961324fc25f4 changes the behavior to tab to the block focus stop first if there are inner children, before proceeding to the inner blocks (in order, so ArrowDown from a previous block).

b2c869ab5d7b0f1d4f6e3a219d4b5bdad5b49e31 adds an overlay on small screen unselected root blocks so that it's required to tap on a nested block wrapper before tapping one of its child blocks.

jasmussen commented:
Thanks for those two fixes, @aduth, it's working great!

I noticed that I can't add anything to the 2nd column in the columns block:

![columns](https://user-images.githubusercontent.com/1204802/40535753-4cbd00c4-600a-11e8-84cf-8cd8bf333a14.gif)

Do you know what might have caused that?

mtias commented:
Looking good! Some thing to address:

* Wide blocks look broken.
* The inner paddings on left and right look a bit large to me.
* With these outlines I have the urge to remove the gray ones when you click on a block. What do you think?
* Sibling inserter is getting too much in the way — can't click the label.

jasmussen commented:
Pushed a few fixes:

- The columns block was broken in a rebase. Fixed that. The appender now works in the right column.
- Made the breadcrumb sit inside, and fade same as the outlines.

> With these outlines I have the urge to remove the gray ones when you click on a block. What do you think?

I worry that will make it a bit fiddly to know what you're interacting with. The biggest challenge is that this will make the block toolbar sit in a fashion that might make it look detached. 

We can definitely explore addressing that — perhaps the block toolbar gets a little triangle arrow that shows what it's connected to. But I think this should be explored separately.

> Sibling inserter is getting too much in the way — can't click the label.

The sibling inserter does need a little further love since the most recent iteration, but at least now you don't accidentally invoke it all the time. But can you clarify why you'd want to click the label? The label doesn't do anything anymore, it's purely informative. 

> The inner paddings on left and right look a bit large to me.

I know. This is because we reserve 28px to the left of all blocks for the side UI, so the clickable area to select the parent actually starts outside of that clickable area:

![wide paddings](https://user-images.githubusercontent.com/1204802/40606109-a0e4ec32-6264-11e8-9e8f-a180ab467d3b.gif)

Right now the extra wide padding is present on all **top level blocks**, but not on **nested blocks**. Some options:

- Make the side UI less wide. Not ideal.
- Try to increase the padding above and below to make it seem more "even"/symmetrical all around.
- Make the wide paddings apply only to blocks that have inner blocks. Paragraphs would look like they do in master, but columns, and in the future other blocks that have children, would get the wide paragraphs. 
- Right now the area below the side UI is used for the _draggable area_ for dragondrop. This extends the height of the block. We could make it so it's really just the up/down arrow that double as the drag handle — then you'd be able to click below the up/down arrows to select the parent block.

jasmussen commented:
P.S. I could use help with a rebase, not sure what's up with the store now.

SuperGeniusZeb commented:
@jasmussen I think making the up/down arrows double as drag handles is a great idea. I think that if you did that, you could also make the block toolbar double as a drag handle as well, providing more than enough space to drag a block from without adding any additional weight to the UI.

Right now, I think the dragging from the side of blocks is kind of confusing. It looks like you are outside the block boundaries, and there is little indicator that the side area is connected to the block. This could be resolved with adding a hover effect to the side of blocks, but I would prefer if the side padding was simply done away with and the arrow buttons and/or block toolbar were made into drag handles.

youknowriad commented:
Quick feedback:

 - Can we drop the line from the in-between inserter, I feel the "plus" button in the middle is enough?
 - The post title width (with outline) is different than the blocks (same for the code editor but this may be a separate issue, I think if we switch the editor mode, things should stay in their exact position)

jasmussen commented:
Good feedback, thank you, and thanks for the rebase help!

> Can we drop the line from the in-between inserter, I feel the "plus" button in the middle is enough?

I agree the sibling inserter needs a little love. The plus in the middle is the only clickable element, which helped make it somewhat more intentional to access, but this isn't working when you are trying to insert above the block you have selected, because the plus sits behind the block toolbar. 

The line indicates where the block will be inserted — perhaps we could color it according to the theme to make it more clear what it means, but I'm not sure if removing it is the solution.

Does it make sense to simply not be able to use the sibling inserter before or after the block that is selected? Does it make sense to require that you simply have no block selected in order to use the sibling inserter? I feel like the answer to both of those are "no", but I'm not sure how to reconcile it otherwise. 

Another solution is to show the plus to the left or right of the sibling inserter, but I tried this in the branch and it conflicted with the side UI. Perhaps we can show it waaay left or waaay right of the block, to not conflict?

> The post title width (with outline) is different than the blocks (same for the code editor but this may be a separate issue, I think if we switch the editor mode, things should stay in their exact position)

This should be fixed in https://github.com/WordPress/gutenberg/pull/6909

youknowriad commented:
I think it's already too crowded to add a line on hover. Maybe the line could only be shown when we hover the button directly. Just my opinion. The issue for me is not related to the selected block.

Also, I personally don't like the "blue" color of the hover outlines, I know it's done on purpose but it feels odd with the rest of the UI

youknowriad commented:
I noticed a small issue when you hover empty paragraph blocks, the inserter on the left should have a white background to hide the outline or something.

jasmussen commented:
Pushed a few changes to address feedback. Before we make more changes, we still need a few answers to https://github.com/WordPress/gutenberg/pull/6773#issuecomment-392464529:

1. Should we remove the draggable area that's vertical so we can make the paddings smaller? ([See also](https://github.com/WordPress/gutenberg/pull/6773#issuecomment-392512051)) This'll require a little rewriting

2. What did you mean, @mtias, when you said "Sibling inserter is getting too much in the way — can't click the label"? — The label isn't clickable anymore.

I'm going to take a stab at the sibling inserter again, see if I can tweak how it works. I think I want to try and move it to the side again, and make the line appear on button hover.

Commit:
Try a refined approach to drag and drop

This commit changes how drag and drop works, so that nested blocks do not have any invisible side affordance for the grabby handle. You can still use the inner padding for dragging, and in a separate enhancement we should make the hover label draggable.

This makes it easier to select the parent block, by allowing you to click the available space below the side UI to select the parent.

Kinda.

There is still an affordance for hovering the side of the block to access the side UI (movers, settings, trash), and this extends the height of 4 stacked buttons on the side. If you hover the area below that, yes you can click the thing below.

This is a balancing act, and I'd like for you to try this. The branch has been thoroughly rebased and squashed prior to this to make it easy to undo this change if it doesn't feel right.

jasmussen commented:
Pushed a change to refine how drag and drop works, inspired by https://github.com/WordPress/gutenberg/pull/6773#issuecomment-392512051:

![side ui](https://user-images.githubusercontent.com/1204802/40657023-a650a32c-6346-11e8-80b8-b54e77f55749.gif)

More details in https://github.com/WordPress/gutenberg/pull/6773/commits/ced659dbf6d72826be13fc3b50a6aece342ed46b, but the gist is:

- nested blocks do not have any invisible side affordance for the grabby handle
- you can still use the inner padding for dragging
- in a separate enhancement we should make the hover label draggable
- this makes it easier to select the parent block, by allowing you to click the available space below the side UI to select the parent
- there is still an affordance for hovering the side of the block to access the side UI (movers, settings, trash), and this extends the height of 4 stacked buttons on the side. If you hover the area below that, yes you can click the thing below.

This is a balancing act, and I'd like for you to try this. The branch has been thoroughly rebased and squashed prior to this to make it easy to undo this change if it doesn't feel right.

SuperGeniusZeb commented:
@jasmussen
I just tried out the latest version of this branch, and I think this is a good improvement. Already the editor feels a lot nicer to use.

One somewhat confusing thing I noticed, though, is that when hovering close below the side controls, the cursor becomes the drag cursor, but you cannot drag either the currently-selected nor the parent block from that spot. You should be able to select the parent block from anywhere on the side of a block except the controls, though the area close below the controls should still cause the controls to show up like it does now.

Also, the side controls, block toolbar, and hover title should definitely all become drag handles. The block toolbar, in particular, would allow a large drag handle to always be on-screen for the selected block (when you are not using the top block toolbar, anyway).

>Does it make sense to simply not be able to use the sibling inserter before or after the block that is selected? Does it make sense to require that you simply have no block selected in order to use the sibling inserter? I feel like the answer to both of those are "no", but I'm not sure how to reconcile it otherwise.

I think having the sibling/in-between inserter not show up next to the currently-selected block would be really bad for user experience. It would certainly confuse me and remove options.

>Another solution is to show the plus to the left or right of the sibling inserter, but I tried this in the branch and it conflicted with the side UI. Perhaps we can show it waaay left or waaay right of the block, to not conflict?

I think showing it to the left or right would not be ideal. Showing it in the center makes it more clear where the block is being inserted and just feels right to me. (Notably, the Divi Visual Builder shows insert icons in the center below modules, rows, and sections much like the current in-between inserter in Gutenberg.) Additionally, showing it to the far left or right would disconnect the inserter from where it is actually inserting.

The blue color on hover does not bother me, and I like how it draws attention to the block you are hovering over. A dark gray color may work just as well, though. I am not sure.

Something interesting I noticed is that when I inserted a Columns block and typed content into it, there was no placeholder/appender at the bottom of either of the columns. I am not sure if this is intentional or not, but I like it, since it removes the extra space at the bottom of nested blocks. There is one important thing that should be changed, though. The in-between inserter should show up at the bottom of each column to allow you to insert more blocks easily. See also: #6834.

I also noticed that I was no longer able to drag the block from the arrow controls when they were disabled. Previously, if one of the arrow controls was grayed out, I could use it as a drag handle. I am guessing this is a bug?

@youknowriad

>I think it's already too crowded to add a line on hover. Maybe the line could only be shown when we hover the button directly.

I think this could work and may help lighten the UI a bit without removing the usefulness of the line. Again, this reminds me of what the Divi Visual Builder does.

>I noticed a small issue when you hover empty paragraph blocks, the inserter on the left should have a white background to hide the outline or something.

I noticed this too, but I forgot about it. Thanks for pointing that out.

Another thing I noticed is that in a custom block I made with nesting, empty/placeholder Paragaph blocks were showing the inserter icon far off to the left, where the icon was on top of the border of the parent block. It seemed strange, since in the Columns block, nested empty/placeholder Paragraph blocks show an inserter icon within their own borders and to the right. I am guessing that is a bug?

![image](https://user-images.githubusercontent.com/19592990/40665872-fc355daa-6323-11e8-9eac-6389ef033b9b.png)

SuperGeniusZeb commented:
@jasmussen This branch is already a good improvement from master, in my opinion, so I am completely fine with getting this merged in as-is after fixing its bugs and focusing on the inserter and further improvements in other issues/PRs.

Also, I had not seen that Squarespace inserter before. That actually looks quite nice, though unlike Gutenberg, Squarespace does not have arrow controls that could be squished up close to those controls or cause the inserter to have to be displaced.

karmatosed commented:
+1 to this being merged and iterating, it's big and an improvement so we should merge now.

jasmussen commented:
Cool, thanks. I'll take a look at the remaining issues mentioned, try to polish a bit, and hopefully we can merge tomorrow.

Then we'll look at sibling inserter and dragon drop separately.

jasmussen commented:
I will for sure look to refine the hover label, and increase paddings. Especially in light of making this be the drag handle. However if I don't do that in this PR, I will ticket and have done separately.

Commit:
Fix the empty appender in more than columns block

The empty appender, when on mobile, or in a nested block, shows a reduced version. Previously this was only in the columns block, now it's any block that uses nested children.

Commit:
Tune hover label again.

jasmussen commented:
![new blue](https://user-images.githubusercontent.com/1204802/40718492-84351186-6410-11e8-8f4d-2d45b47b1c57.gif)

New hover label

Commit:
Fix rebase issue.

jasmussen commented:
Okay, from an experience and CSS point of view, this is now ready.

This PR brings a new way of selecting blocks, including parent blocks, both for mouse, keyboard and mobile users. It refactors a lot of variables and refines a bunch of styles in order to get there, and the hover label has been re-thoughts.

There's one last thing that I'd like help with, @gziolo or @youknowriad or @jorgefilipecosta if you have time — and that is hiding the hover outline and hover label from an empty paragraph block:

<img width="863" alt="screen shot 2018-05-30 at 13 52 18" src="https://user-images.githubusercontent.com/1204802/40719415-db52f6c4-6413-11e8-838a-069bada5727c.png">

Other than that, I think it's ready for a final review.

youknowriad committed:
hide outlines from empty default blocks

notnownikki commented:
Just one little thing I found, on a classic block the hover label is obscured by the classic block header.

![overlap](https://user-images.githubusercontent.com/3314354/40723717-68957ec8-6417-11e8-90d6-a70351e66b3b.png)

Also, I feel like the blue doesn't quite fit in with the color scheme... but no one ever asked me to coordinate their outfit colors, so perhaps don't pay too much attention to that :)

jasmussen commented:
Great catches. Will fix the classic block, it looks wrong entirely, and using a theme color for hover and label might be a fun splash of color.

jasmussen committed:
Fix paddings and hover breadcrumb on classic block.

jasmussen committed:
Fade the title block border.

jasmussen commented:
Changing the color of the hovered block and the label to match the admin theme is easy to do, but I would like to do it separately to keep this PR from growing further.

notnownikki commented:
Looking good!

mtias commented:
I'm seeing the spacer block missing controls on this branch:

![image](https://user-images.githubusercontent.com/548849/40733229-6d4117b6-6435-11e8-8c20-78b9a5d9e8e6.png)

aduth commented:
>I'm seeing the spacer block missing controls on this branch:

Noting as unrelated to these changes. See #7024, https://github.com/WordPress/gutenberg/pull/6922#issuecomment-393236009.

aduth commented on editor/components/block-mover/style.scss:
```diff
@@ -47,15 +47,16 @@
+			&:hover,
+			&:active,
+			&:focus {
+				z-index: z-index( '.editor-block-mover__control' );
```
Knowing `z-index` to be a source of maintenance headaches, I might like to see some inline comment either here or in `z-index.scss` explaining why it's needed. As it stands, I have no idea why we need it, even at this moment in time, let alone my future self.

jasmussen replied:
👍👍 Adding a comment.

aduth commented on editor/components/block-toolbar/style.scss:
```diff
@@ -13,7 +13,11 @@
+	}
+
+	// Only show the left border if the Switcher is present
+	.editor-block-switcher + div > .components-toolbar {
```
This seems very prone to future breakage, particularly on the `+ div` sibling part, since the `div` it refers to exists only for adding a left border in a set of toolbar controls. This has no reason for needing to be a `div` specifically, and could otherwise be refactored to a solution which no longer requires the `div` wrapper at all.

Instead, I'd like to see one of:

- Adding a `border-right` to the `Toolbar` of the `EditorBlockSwitcher`
- See what we can do about removing [this "seemingly redundant wrapper" `div`](https://github.com/WordPress/gutenberg/blob/6728906116c12b5865e66b32b634d2b300136334/components/dropdown/index.js#L75-L79) and change selector to apply `border-left` on `.editor-block-toolbar .components-toolbar:not( :first-child )`
   - Alas, we can't use [`:first-of-type`](https://developer.mozilla.org/en-US/docs/Web/CSS/:first-of-type) for anything except an element tag ([reference](https://stackoverflow.com/a/6447072/995445)). Well, maybe [someday](https://stackoverflow.com/a/19474149/995445)).
- Add `hasLeftDivider` as a prop to control / override the existing behavior of `Toolbar` in applying the `.has-left-divider` modifier class

jasmussen replied:
The first idea was a solid bit of advice. Did that:

![border fix](https://user-images.githubusercontent.com/1204802/40828899-3178e5b2-6582-11e8-80bf-bbc0f2c114b1.gif)

aduth reviewed (approved):
Code-wise 👍

Commit:
Address feedback.

Add border to switcher component instead.

Add comment to explain z index.

jasmussen commented:
🤘

iseulde commented:
![drag-label](https://user-images.githubusercontent.com/4710635/40840100-94742ea6-65a5-11e8-9179-0b479b5112c2.gif)

Dragging seems broken here

aduth commented:
@iseulde If you're referring to the block label, dragging has never worked there. I know it's been mentioned a few times as being desirable, though unsure if there's an issue tracking it.

jasmussen commented:
Well, it does _work_ for me — it's just that you have to click and drag and it's not a great experience yet.

It was my intention to create a separate ticket for this, including a ticket to improve the sibling inserter and theme the hover color, per feedback in this thread. But it's been a whirlwind so creating that ticket is still on my "todo". But the gist is I'm going to suppose sort of a rejiggering of dragondrop so it's _only_ the label that becomes the drag handle, and not the sides. But I'm sure lots of discussion and idea refinement will ensue.

jasmussen commented:
Clicking and dragging that area is working for me, though, even if in a way that doesn't give the best visual feedback. Is it not for you?

SuperGeniusZeb commented:
@jasmussen Drag and drop from the hover title is not working for me either. (Tested in Firefox Nightly.) The cursor becomes the drag cursor when I hover over the title, but when I try to click and hold on the hover title, it just selects the block.

jasmussen commented:
It worked on the edge, but I see why there's confusion: it doesn't work on the _whole_ breadcrumb. So yes, it was definitely too early to add the grab handle. I've opened #7083 PR to fix it.

aduth commented on core-blocks/columns/style.scss:
```diff
@@ -1,9 +1,12 @@
 
 	@for $i from 2 through 6 {
-		&.has-#{ $i }-columns {
+		&.has-#{ $i }-columns .editor-inner-blocks {
```
This change seems to have inadvertently caused #7127, where the stylesheet is used for both the editor and front-end, but the markup generated from the front-end will not have `.editor-inner-blocks`.

jasmussen replied:
Looking.
